### DISCUSSION TITLE
Questions about equivocation implementation in Babe

### DISCUSSION THREAD

**Discussion by nazar-pc** [paritytech/substrate#9327] [Q&A] [answered]
I'm working on equivocation support for Spartan as part of W3F grant and faced difficulties.

We have started with Babe as foundation (it shares quite a few similarities due to slots usage), but as out protocol is based on proof of capacity, we don't have authorities.

When time came to implement equivocation I took what Babe, pallets staking and offences and modified those to not have reporters or report authoring.

Once things compiled and ran, I modified [`check_equivocation` in `cs_consensus_slots`](https://github.com/subspace/substrate/blob/93b231e79f5b4e551c34234e89fa4a2e5e9c1510/client/consensus/slots/src/aux_schema.rs#L53-L134) to always report fake equivocation just to test how things work and found that it actually fails to submit transaction with "Could not find an unsigned validator for the unsigned transaction" (internally) error:
```
2021-07-12 00:05:23  Slot author Public(2ee367fe3543a86ded8ab4e27bc106b24fc20698206493edc102b6275ee6ef15 (5D8BcYTP...)) is equivocating at slot 1626037523 with headers 0x463e8cdc0348dab5590808c7af9ac2a9d40b4eee99ad22d4cf5a7504bd72151a and 0x463e8cdc0348dab5590808c7af9ac2a9d40b4eee99ad22d4cf5a7504bd72151a    
2021-07-12 00:05:23  (offchain call) Error submitting a transaction to the pool: Pool(UnknownTransaction(UnknownTransaction::NoUnsignedValidator))    
2021-07-12 00:05:23  Error submitting equivocation report: ()    
```

From this there are actually 2 questions:
1) What that error means exactly (it supposedly works in Babe and I didn't change the code path to diverge from it too much)?
2) Why does `BabeVerifier::check_and_report_equivocation()` needs to go all the way through runtime to do what seems like addition of unsigned transaction to transaction pool?

I'm sorry for a bunch of info, but hopefully it makes sense. Our fork of substrate's master branch if from April 11th, so a few things has changed in master already :disappointed:

**Comment by xlc** [comment, marked as answer]
https://github.com/paritytech/substrate/blob/efd54879ecbf8aac41fe809d31c7215ea3101241/bin/node/runtime/src/lib.rs#L1170

You missing `ValidateUnsigned`

**Reply by nazar-pc** [reply]
Wow, that does help indeed!

Do you happen to know why so much indirection is needed to insert it into transaction pool?

**Reply by xlc** [reply]
https://github.com/paritytech/substrate/issues/8084

**Reply by nazar-pc** [reply]
Hm, I do not see how it is relevant here :thinking:
